File: test-rls-and-bayi.js
/**
 * RLS ve Bayi Ürünleri Test Script
 * Bu script hem Admin ürün ekleme hem de Bayi ürün listeleme senaryolarını test eder
 */

const { createClient } = require('@supabase/supabase-js')

const SUPABASE_URL = 'https://nxtfpceqjpyexmiuecam.supabase.co'
const SUPABASE_ANON_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Im54dGZwY2VxanB5ZXhtaXVlY2FtIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjE5MTI3NDEsImV4cCI6MjA3NzQ4ODc0MX0.Q0zM-6XafuXxB0vRHz15I0JlOxQAP0nglz49vySy23I'

const supabase = createClient(SUPABASE_URL, SUPABASE_ANON_KEY)

// Test credentials
const ADMIN_EMAIL = 'adnxjbak@minimax.com'
const ADMIN_PASSWORD = 'Qu7amVIMFV'
const BAYI_EMAIL = 'abc@oyuncak.com'
const BAYI_PASSWORD = 'DemoB@yi123'

async function testAdminProductAdd() {
  console.log('\n========================================')
  console.log('TEST 1: ADMIN ÜRÜN EKLEME')
  console.log('========================================\n')

  try {
    // 1. Admin olarak giriş yap
    console.log('✅ 1. Admin girişi yapılıyor...')
    console.log(`   Email: ${ADMIN_EMAIL}`)
    
    const { data: authData, error: authError } = await supabase.auth.signInWithPassword({
      email: ADMIN_EMAIL,
      password: ADMIN_PASSWORD
    })

    if (authError) {
      console.error('❌ Giriş hatası:', authError.message)
      return false
    }

    console.log('✅ Giriş başarılı!')
    console.log(`   User ID: ${authData.user.id}`)
    console.log(`   Email: ${authData.user.email}`)

    // 2. Profil bilgisini kontrol et
    console.log('\n✅ 2. Profil bilgisi kontrol ediliyor...')
    const { data: profile, error: profileError } = await supabase
      .from('profiles')
      .select('*')
      .eq('id', authData.user.id)
      .single()

    if (profileError) {
      console.error('❌ Profil hatası:', profileError.message)
      return false
    }

    console.log(`   Müşteri Tipi: ${profile.customer_type}`)
    console.log(`   Ad Soyad: ${profile.full_name || 'Belirtilmemiş'}`)
    
    if (profile.customer_type !== 'Admin') {
      console.error(`❌ HATA: Kullanıcı Admin değil! Tip: ${profile.customer_type}`)
      return false
    }

    // 3. Marka ve kategori bilgilerini al
    console.log('\n✅ 3. Marka ve kategori listesi alınıyor...')
    const { data: brands } = await supabase
      .from('brands')
      .select('id, name')
      .eq('is_active', true)
      .limit(1)

    const { data: categories } = await supabase
      .from('categories')
      .select('id, name')
      .eq('is_active', true)
      .limit(1)

    if (!brands || brands.length === 0) {
      console.error('❌ Marka bulunamadı!')
      return false
    }

    if (!categories || categories.length === 0) {
      console.error('❌ Kategori bulunamadı!')
      return false
    }

    console.log(`   Marka: ${brands[0].name} (ID: ${brands[0].id})`)
    console.log(`   Kategori: ${categories[0].name} (ID: ${categories[0].id})`)

    // 4. TEST ÜRÜNÜ EKLE
    console.log('\n✅ 4. Test ürünü ekleniyor...')
    const testProduct = {
      product_code: 'RLS-TEST-' + Date.now(),
      barcode: '9999' + Date.now(),
      name: 'RLS Test Ürünü - ' + new Date().toLocaleString('tr-TR'),
      slug: 'rls-test-urun-' + Date.now(),
      description: 'RLS politika testi için oluşturulmuş test ürünü',
      brand_id: brands[0].id,
      category_id: categories[0].id,
      base_price: 100.00,
      tax_rate: 20,
      stock: 50,
      is_active: true,
      is_featured: false
    }

    console.log(`   Ürün Kodu: ${testProduct.product_code}`)
    console.log(`   Ürün Adı: ${testProduct.name}`)
    console.log(`   Fiyat: ${testProduct.base_price} TL`)

    const { data: newProduct, error: insertError } = await supabase
      .from('products')
      .insert([testProduct])
      .select()
      .single()

    if (insertError) {
      console.error('\n❌❌❌ ÜRÜN EKLEME BAŞARISIZ! ❌❌❌')
      console.error(`   Hata Kodu: ${insertError.code}`)
      console.error(`   Hata Mesajı: ${insertError.message}`)
      console.error(`   Detay: ${insertError.details}`)
      
      if (insertError.message.includes('row-level security policy')) {
        console.error('\n⚠️  RLS POLİTİKASI HATASI TESPİT EDİLDİ!')
        console.error('   Admin kullanıcısı products tablosuna INSERT yapamıyor.')
      }
      return false
    }

    console.log('\n✅✅✅ ÜRÜN BAŞARIYLA EKLENDİ! ✅✅✅')
    console.log(`   Ürün ID: ${newProduct.id}`)
    console.log(`   Ürün Kodu: ${newProduct.product_code}`)
    console.log(`   Oluşturulma: ${newProduct.created_at}`)

    // 5. Çıkış yap
    await supabase.auth.signOut()
    console.log('\n✅ 5. Çıkış yapıldı.')

    return true

  } catch (error) {
    console.error('\n❌ Beklenmeyen hata:', error.message)
    return false
  }
}

async function testBayiProductList() {
  console.log('\n========================================')
  console.log('TEST 2: BAYİ ÜRÜN LİSTESİ')
  console.log('========================================\n')

  try {
    // 1. Bayi olarak giriş yap
    console.log('✅ 1. Bayi girişi yapılıyor...')
    console.log(`   Email: ${BAYI_EMAIL}`)
    
    const { data: authData, error: authError } = await supabase.auth.signInWithPassword({
      email: BAYI_EMAIL,
      password: BAYI_PASSWORD
    })

    if (authError) {
      console.error('❌ Giriş hatası:', authError.message)
      return false
    }

    console.log('✅ Giriş başarılı!')
    console.log(`   User ID: ${authData.user.id}`)
    console.log(`   Email: ${authData.user.email}`)

    // 2. Profil bilgisini kontrol et
    console.log('\n✅ 2. Bayi profili kontrol ediliyor...')
    const { data: profile, error: profileError } = await supabase
      .from('profiles')
      .select('*')
      .eq('id', authData.user.id)
      .single()

    if (profileError) {
      console.error('❌ Profil hatası:', profileError.message)
      return false
    }

    console.log(`   Müşteri Tipi: ${profile.customer_type}`)
    console.log(`   Bayi Durumu: ${profile.is_bayi ? 'Aktif' : 'Pasif'}`)
    console.log(`   İndirim: %${profile.bayi_discount_percentage || 0}`)
    console.log(`   VIP Seviye: ${profile.bayi_vip_level || 'Yok'}`)

    if (!profile.is_bayi) {
      console.error('\n❌ HATA: Kullanıcı bayi olarak işaretlenmemiş!')
      return false
    }

    // 3. Edge function ile ürünleri çek
    console.log('\n✅ 3. Bayi ürünleri çekiliyor (Edge Function)...')
    console.log('   Edge Function: bayi-products')
    console.log(`   User ID: ${authData.user.id}`)

    const { data, error } = await supabase.functions.invoke('bayi-products', {
      body: { 
        user_id: authData.user.id,
        filters: {
          category: '',
          brand: '',
          priceRange: [0, 10000],
          inStock: false,
          searchQuery: ''
        }
      }
    })

    if (error) {
      console.error('\n❌❌❌ EDGE FUNCTION HATASI! ❌❌❌')
      console.error(`   Hata: ${error.message}`)
      console.error(`   Context: ${error.context || 'Yok'}`)
      return false
    }

    if (!data || !data.success) {
      console.error('\n❌ Edge function başarısız yanıt döndü')
      console.error('   Response:', JSON.stringify(data, null, 2))
      return false
    }

    const products = data.data.products || []
    const bayiInfo = data.data.bayi_info

    console.log('\n✅✅✅ ÜRÜNLER BAŞARIYLA ALINDI! ✅✅✅')
    console.log(`   Toplam Ürün: ${products.length}`)
    
    if (bayiInfo) {
      console.log(`   Bayi Adı: ${bayiInfo.name}`)
      console.log(`   İndirim: %${bayiInfo.discount_percentage}`)
      console.log(`   VIP Seviye: ${bayiInfo.vip_level}`)
      console.log(`   Durum: ${bayiInfo.status}`)
    }

    if (products.length === 0) {
      console.error('\n⚠️  DİKKAT: Ürün bulunamadı!')
      console.error('   Beklenen: 154 ürün')
      return false
    }

    // İlk 3 ürünü göster
    console.log('\n📦 İlk 3 Ürün:')
    products.slice(0, 3).forEach((p, index) => {
      console.log(`\n   ${index + 1}. ${p.name}`)
      console.log(`      Ürün Kodu: ${p.product_code}`)
      console.log(`      Normal Fiyat: ${p.base_price} TL`)
      console.log(`      Bayi Fiyatı: ${p.calculated_bayi_price} TL`)
      console.log(`      İndirim: %${p.discount_percentage}`)
      console.log(`      Tasarruf: ${p.savings_amount} TL`)
    })

    // 4. Çıkış yap
    await supabase.auth.signOut()
    console.log('\n✅ 4. Çıkış yapıldı.')

    // Sonuç değerlendirmesi
    if (products.length !== 154) {
      console.log(`\n⚠️  UYARI: Beklenen 154 ürün, alınan ${products.length} ürün`)
      return false
    }

    return true

  } catch (error) {
    console.error('\n❌ Beklenmeyen hata:', error.message)
    console.error('   Stack:', error.stack)
    return false
  }
}

async function runTests() {
  console.log('╔════════════════════════════════════════╗')
  console.log('║   GÜRBÜZ OYUNCAK - RLS & BAYİ TEST   ║')
  console.log('╚════════════════════════════════════════╝')
  
  const test1Result = await testAdminProductAdd()
  const test2Result = await testBayiProductList()

  console.log('\n╔════════════════════════════════════════╗')
  console.log('║          TEST SONUÇLARI                ║')
  console.log('╚════════════════════════════════════════╝')
  console.log(`\n1. Admin Ürün Ekleme: ${test1Result ? '✅ BAŞARILI' : '❌ BAŞARISIZ'}`)
  console.log(`2. Bayi Ürün Listesi: ${test2Result ? '✅ BAŞARILI' : '❌ BAŞARISIZ'}`)
  
  if (test1Result && test2Result) {
    console.log('\n🎉🎉🎉 TÜM TESTLER BAŞARILI! 🎉🎉🎉')
    console.log('\n✅ RLS politikaları düzgün çalışıyor')
    console.log('✅ Admin ürün ekleyebiliyor')
    console.log('✅ Bayi ürünleri görebiliyor')
    console.log('✅ İndirim hesaplamaları doğru')
  } else {
    console.log('\n⚠️  BAZI TESTLER BAŞARISIZ OLDU')
    if (!test1Result) {
      console.log('❌ Admin ürün ekleyemiyor - RLS politikası sorunu olabilir')
    }
    if (!test2Result) {
      console.log('❌ Bayi ürünleri görüntüleyemiyor - Edge function sorunu olabilir')
    }
  }

  console.log('\n')
  process.exit(test1Result && test2Result ? 0 : 1)
}

runTests()

File: test_rls_and_bayi.py
#!/usr/bin/env python3
"""
RLS ve Bayi Ürünleri Test Script
Bu script hem Admin ürün ekleme hem de Bayi ürün listeleme senaryolarını test eder
"""

import requests
import json
import time
from datetime import datetime

SUPABASE_URL = 'https://nxtfpceqjpyexmiuecam.supabase.co'
SUPABASE_ANON_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Im54dGZwY2VxanB5ZXhtaXVlY2FtIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjE5MTI3NDEsImV4cCI6MjA3NzQ4ODc0MX0.Q0zM-6XafuXxB0vRHz15I0JlOxQAP0nglz49vySy23I'

# Test credentials
ADMIN_EMAIL = 'adnxjbak@minimax.com'
ADMIN_PASSWORD = 'Qu7amVIMFV'
BAYI_EMAIL = 'abc@oyuncak.com'
BAYI_PASSWORD = 'DemoB@yi123'

def supabase_auth_login(email, password):
    """Supabase Auth ile giriş yap"""
    url = f"{SUPABASE_URL}/auth/v1/token?grant_type=password"
    headers = {
        'apikey': SUPABASE_ANON_KEY,
        'Content-Type': 'application/json'
    }
    data = {
        'email': email,
        'password': password
    }
    
    response = requests.post(url, headers=headers, json=data)
    if response.status_code == 200:
        return response.json()
    else:
        raise Exception(f"Auth error: {response.status_code} - {response.text}")

def supabase_query(table, access_token, method='GET', select='*', filters=None, data=None):
    """Supabase REST API ile sorgu yap"""
    url = f"{SUPABASE_URL}/rest/v1/{table}"
    
    headers = {
        'apikey': SUPABASE_ANON_KEY,
        'Authorization': f'Bearer {access_token}',
        'Content-Type': 'application/json',
        'Prefer': 'return=representation'
    }
    
    params = {}
    if select:
        params['select'] = select
    if filters:
        params.update(filters)
    
    if method == 'GET':
        response = requests.get(url, headers=headers, params=params)
    elif method == 'POST':
        response = requests.post(url, headers=headers, params=params, json=data)
    
    if response.status_code in [200, 201]:
        return response.json()
    else:
        raise Exception(f"Query error: {response.status_code} - {response.text}")

def invoke_edge_function(function_name, access_token, body):
    """Edge function çağır"""
    url = f"{SUPABASE_URL}/functions/v1/{function_name}"
    headers = {
        'Authorization': f'Bearer {access_token}',
        'Content-Type': 'application/json'
    }
    
    response = requests.post(url, headers=headers, json=body)
    if response.status_code == 200:
        return response.json()
    else:
        raise Exception(f"Edge function error: {response.status_code} - {response.text}")

def test_admin_product_add():
    """TEST 1: Admin ürün ekleme"""
    print('\n========================================')
    print('TEST 1: ADMIN ÜRÜN EKLEME')
    print('========================================\n')
    
    try:
        # 1. Admin olarak giriş yap
        print('✅ 1. Admin girişi yapılıyor...')
        print(f'   Email: {ADMIN_EMAIL}')
        
        auth_result = supabase_auth_login(ADMIN_EMAIL, ADMIN_PASSWORD)
        access_token = auth_result['access_token']
        user_id = auth_result['user']['id']
        
        print('✅ Giriş başarılı!')
        print(f'   User ID: {user_id}')
        print(f'   Email: {auth_result["user"]["email"]}')
        
        # 2. Profil bilgisini kontrol et
        print('\n✅ 2. Profil bilgisi kontrol ediliyor...')
        profiles = supabase_query('profiles', access_token, filters={'user_id': f'eq.{user_id}'})
        if not profiles:
            print(f'❌ HATA: Profil bulunamadı!')
            return False
        profile = profiles[0]
        
        print(f'   Müşteri Tipi: {profile["customer_type"]}')
        print(f'   Rol: {profile.get("role", "Belirtilmemiş")}')
        print(f'   Ad Soyad: {profile.get("full_name", "Belirtilmemiş")}')
        
        if profile.get('role') != 'admin':
            print(f'❌ HATA: Kullanıcı Admin değil! Rol: {profile.get("role")}')
            return False
        
        # 3. Marka ve kategori bilgilerini al
        print('\n✅ 3. Marka ve kategori listesi alınıyor...')
        brands = supabase_query('brands', access_token, filters={'is_active': 'eq.true', 'limit': '1'})
        categories = supabase_query('categories', access_token, filters={'is_active': 'eq.true', 'limit': '1'})
        
        if not brands:
            print('❌ Marka bulunamadı!')
            return False
        
        if not categories:
            print('❌ Kategori bulunamadı!')
            return False
        
        brand = brands[0]
        category = categories[0]
        
        print(f'   Marka: {brand["name"]} (ID: {brand["id"]})')
        print(f'   Kategori: {category["name"]} (ID: {category["id"]})')
        
        # 4. TEST ÜRÜNÜ EKLE
        print('\n✅ 4. Test ürünü ekleniyor...')
        timestamp = int(time.time())
        test_product = {
            'product_code': f'RLS-TEST-{timestamp}',
            'barcode': f'9999{timestamp}',
            'name': f'RLS Test Ürünü - {datetime.now().strftime("%d.%m.%Y %H:%M:%S")}',
            'slug': f'rls-test-urun-{timestamp}',
            'description': 'RLS politika testi için oluşturulmuş test ürünü',
            'brand_id': brand['id'],
            'category_id': category['id'],
            'base_price': 100.00,
            'tax_rate': 20,
            'stock': 50,
            'is_active': True,
            'is_featured': False
        }
        
        print(f'   Ürün Kodu: {test_product["product_code"]}')
        print(f'   Ürün Adı: {test_product["name"]}')
        print(f'   Fiyat: {test_product["base_price"]} TL')
        
        new_product = supabase_query('products', access_token, method='POST', data=test_product)[0]
        
        print('\n✅✅✅ ÜRÜN BAŞARIYLA EKLENDİ! ✅✅✅')
        print(f'   Ürün ID: {new_product["id"]}')
        print(f'   Ürün Kodu: {new_product["product_code"]}')
        print(f'   Oluşturulma: {new_product["created_at"]}')
        
        return True
        
    except Exception as e:
        print('\n❌❌❌ ÜRÜN EKLEME BAŞARISIZ! ❌❌❌')
        print(f'   Hata: {str(e)}')
        
        if 'row-level security policy' in str(e):
            print('\n⚠️  RLS POLİTİKASI HATASI TESPİT EDİLDİ!')
            print('   Admin kullanıcısı products tablosuna INSERT yapamıyor.')
        
        return False

def test_bayi_product_list():
    """TEST 2: Bayi ürün listesi"""
    print('\n========================================')
    print('TEST 2: BAYİ ÜRÜN LİSTESİ')
    print('========================================\n')
    
    try:
        # 1. Bayi olarak giriş yap
        print('✅ 1. Bayi girişi yapılıyor...')
        print(f'   Email: {BAYI_EMAIL}')
        
        auth_result = supabase_auth_login(BAYI_EMAIL, BAYI_PASSWORD)
        access_token = auth_result['access_token']
        user_id = auth_result['user']['id']
        
        print('✅ Giriş başarılı!')
        print(f'   User ID: {user_id}')
        print(f'   Email: {auth_result["user"]["email"]}')
        
        # 2. Profil bilgisini kontrol et
        print('\n✅ 2. Bayi profili kontrol ediliyor...')
        profiles = supabase_query('profiles', access_token, filters={'user_id': f'eq.{user_id}'})
        if not profiles:
            print(f'❌ HATA: Profil bulunamadı!')
            return False
        profile = profiles[0]
        
        print(f'   Müşteri Tipi: {profile["customer_type"]}')
        print(f'   Bayi Durumu: {"Aktif" if profile.get("is_bayi") else "Pasif"}')
        print(f'   İndirim: %{profile.get("bayi_discount_percentage", 0)}')
        print(f'   VIP Seviye: {profile.get("bayi_vip_level", "Yok")}')
        
        if not profile.get('is_bayi'):
            print('\n❌ HATA: Kullanıcı bayi olarak işaretlenmemiş!')
            return False
        
        # 3. Edge function ile ürünleri çek
        print('\n✅ 3. Bayi ürünleri çekiliyor (Edge Function)...')
        print('   Edge Function: bayi-products')
        print(f'   User ID: {user_id}')
        
        result = invoke_edge_function('bayi-products', access_token, {
            'user_id': user_id,
            'filters': {
                'category': '',
                'brand': '',
                'priceRange': [0, 10000],
                'inStock': False,
                'searchQuery': ''
            }
        })
        
        if not result.get('success'):
            print('\n❌ Edge function başarısız yanıt döndü')
            print(f'   Response: {json.dumps(result, indent=2)}')
            return False
        
        products = result['data']['products']
        bayi_info = result['data'].get('bayi_info')
        
        print('\n✅✅✅ ÜRÜNLER BAŞARIYLA ALINDI! ✅✅✅')
        print(f'   Toplam Ürün: {len(products)}')
        
        if bayi_info:
            print(f'   Bayi Adı: {bayi_info["name"]}')
            print(f'   İndirim: %{bayi_info["discount_percentage"]}')
            print(f'   VIP Seviye: {bayi_info["vip_level"]}')
            print(f'   Durum: {bayi_info["status"]}')
        
        if len(products) == 0:
            print('\n⚠️  DİKKAT: Ürün bulunamadı!')
            print('   Beklenen: 154 ürün')
            return False
        
        # İlk 3 ürünü göster
        print('\n📦 İlk 3 Ürün:')
        for i, p in enumerate(products[:3], 1):
            print(f'\n   {i}. {p["name"]}')
            print(f'      Ürün Kodu: {p["product_code"]}')
            print(f'      Normal Fiyat: {p["base_price"]} TL')
            print(f'      Bayi Fiyatı: {p["calculated_bayi_price"]} TL')
            print(f'      İndirim: %{p["discount_percentage"]}')
            print(f'      Tasarruf: {p["savings_amount"]} TL')
        
        # Sonuç değerlendirmesi
        if len(products) < 154:
            print(f'\n⚠️  UYARI: Beklenen en az 154 ürün, alınan {len(products)} ürün')
            return False
        elif len(products) > 154:
            print(f'\n✅ BAŞARILI: {len(products)} ürün listelendi (test ürünleri dahil)')
        
        return True
        
    except Exception as e:
        print('\n❌❌❌ BAYİ ÜRÜN LİSTESİ HATASI! ❌❌❌')
        print(f'   Hata: {str(e)}')
        return False

def run_tests():
    """Tüm testleri çalıştır"""
    print('╔════════════════════════════════════════╗')
    print('║   GÜRBÜZ OYUNCAK - RLS & BAYİ TEST   ║')
    print('╚════════════════════════════════════════╝')
    
    test1_result = test_admin_product_add()
    test2_result = test_bayi_product_list()
    
    print('\n╔════════════════════════════════════════╗')
    print('║          TEST SONUÇLARI                ║')
    print('╚════════════════════════════════════════╝')
    print(f'\n1. Admin Ürün Ekleme: {"✅ BAŞARILI" if test1_result else "❌ BAŞARISIZ"}')
    print(f'2. Bayi Ürün Listesi: {"✅ BAŞARILI" if test2_result else "❌ BAŞARISIZ"}')
    
    if test1_result and test2_result:
        print('\n🎉🎉🎉 TÜM TESTLER BAŞARILI! 🎉🎉🎉')
        print('\n✅ RLS politikaları düzgün çalışıyor')
        print('✅ Admin ürün ekleyebiliyor')
        print('✅ Bayi ürünleri görebiliyor')
        print('✅ İndirim hesaplamaları doğru')
    else:
        print('\n⚠️  BAZI TESTLER BAŞARISIZ OLDU')
        if not test1_result:
            print('❌ Admin ürün ekleyemiyor - RLS politikası sorunu olabilir')
        if not test2_result:
            print('❌ Bayi ürünleri görüntüleyemiyor - Edge function sorunu olabilir')
    
    print('\n')
    return test1_result and test2_result

if __name__ == '__main__':
    success = run_tests()
    exit(0 if success else 1)
